### ISSUE TITLE
CUDA configuration is not correct/accurate

### ISSUE THREAD

**Issue by archenroot** [colmap/colmap#145]
Hi, first thanks for this cool stuff here! I am on Gentoo system with GCC 6.3.2 and CUDA 8.0 Toolkit and found following configuration:
```
    list(APPEND CUDA_NVCC_FLAGS "-gencode;arch=compute_20,code=sm_20")
    list(APPEND CUDA_NVCC_FLAGS "-gencode;arch=compute_30,code=sm_30")
    list(APPEND CUDA_NVCC_FLAGS "-gencode;arch=compute_35,code=sm_35")

    if(NOT CUDA_VERSION VERSION_LESS "6.5") # >= 6.5
        list(APPEND CUDA_NVCC_FLAGS "-gencode;arch=compute_37,code=sm_37")
        list(APPEND CUDA_NVCC_FLAGS "-gencode;arch=compute_50,code=sm_50")
    endif()

    if(NOT CUDA_VERSION VERSION_LESS "7.0") # >= 7.0
        list(APPEND CUDA_NVCC_FLAGS "-gencode;arch=compute_52,code=sm_52")
    endif()

    if(NOT CUDA_VERSION VERSION_LESS "8.0") # >= 8.0
        list(APPEND CUDA_NVCC_FLAGS "-gencode;arch=compute_60,code=sm_60")
        list(APPEND CUDA_NVCC_FLAGS "-gencode;arch=compute_61,code=sm_61")
    endif()
```
It is overkill to compile more architectures than you actually have. I am working on Gentoo packages how to handle this from within system configuration with USE_EXPAND variable(Gentoo specific only). 

But, In meanwhile I found also very nice solution used by OpenCV project which includes CMAKE CUDA selection/detection, but there was issue, see details here:
https://github.com/opencv/opencv/issues/8781

At the end of issue above is new latest proper CUDA handling built in CMAKE. 

The idea is not to compile for architectures you don't have to lower compile time (it matters on my system :-) ).

**Comment by ahojnnes** [comment]
It is actually not overkill if you distribute the model for different platforms. That is why it is only activated in `Release` mode and not in e.g. `RelWithDeb`. I will add an option to disable this. Thanks for the feedback!

**Comment by archenroot** [comment]
Make sense.

**Comment by archenroot** [comment]
As I suggested, you could utilize:
https://gitlab.kitware.com/cmake/cmake/blob/v3.7.0/Modules/FindCUDA/select_compute_arch.cmake#L45

**Comment by ahojnnes** [comment]
This seems like a good option but CMake 3.7.x is rather new and not available on systems like Ubuntu 14.04 which I still support. I will check the license of the file and see whether I can just copy the code and ship it with Colmap. Thanks for the link.

**Comment by ahojnnes** [comment]
This should be fixed now. By default, only the current architecture is compiled for. You can enable multi-architecture compilation with ``cmake .. -DCUDA_MULTI_ARCH=ON``. Thanks for the feedback.
